Plain cmake

kimgr opened the pull request:
It looks like Debian packaging for Clang is fixed now, so simplify CMakeLists.txt and documentation.

This PR is mostly to see if Travis picks up the change and builds correctly.

kimgr commented:
This now passes our internal infrastructure, so it would be interesting to hear from external users -- @nilsdeppe, @TheErk could you try this branch in your respective environments? The relevant CMake command is `cmake -DCMAKE_PREFIX_PATH=/usr/lib/llvm-7 -G Ninja ../include-what-you-use`.

Out-of-tree is the way to go, in-tree is no longer supported.

Does anyone have a Windows machine to test this on?

kimgr commented:
Tested on FreeBSD and found a missing detail. Re-pushed.

jru commented:
I am a bit new to this, but why are we getting rid of in-tree builds? Is it
because the don't work for some platforms?


Btw, I can test it on windows today.

kimgr commented:
In-tree builds have historically worked better than out-of-tree, but maintaining both is... delicate.

The in-tree configuration also requires that rogue edit of `clang/tools/CMakeLists.txt`, which is annoying and error-prone.

Since the out-of-tree build can now (fingers crossed!) be configured against a local LLVM/Clang build tree, the in-tree build is no longer pulling its weight, IMHO.

jru commented:
Hi,

I agree having to add the IWYU directory in the clang/tools CMakeLists.txt is hacky, but it was really convenient to pull the latest LLVM/Clang and have IWYU build incrementally from there in no time. I honestly used it all the time for development.

Right now, I will have to build everything and install it so that IWYU can build. It would be nicer if there were prebuilt up-to-date snapshots for windows, but that's not normally the case (the latest right now is from April), so I have to go from source.

More importantly, as far as I tested so far, your cleaned up version works great also for in-tree, by just excluding the find_package stuff from it. Like this:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 36890a5..22c8e7e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -7,6 +7,9 @@ endif()

 project(include-what-you-use)

+if( CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR )
+  message(STATUS "IWYU: out-of-tree configuration")
+
   find_package(LLVM CONFIG REQUIRED)
   find_package(Clang CONFIG REQUIRED)

   message(STATUS "IWYU: configuring for LLVM ${LLVM_VERSION}...")

   list(APPEND CMAKE_MODULE_PATH ${LLVM_DIR})
   include(HandleLLVMOptions)

+else()
+  message(STATUS "IWYU: in-tree configuration")
+endif()
+
 add_definitions(${LLVM_DEFINITIONS})
 include_directories(
   ${LLVM_INCLUDE_DIRS}

```

That said, more testing is necessary. I'll give out-of-tree a try and keep you up to date!

kimgr commented:
> Right now, I will have to build everything and install it so that IWYU can build.

Note that you only have to *build* it, not *install* it. You can pass Clang's build root as `CMAKE_PREFIX_PATH` (as mentioned in the revised docs). But I agree it is less convenient, I used to always prefer the in-tree build for the reasons you cite (and because it was the only configuration initially).

I think users of IWYU will always use out-of-tree (from a release branch/tag) against a local install of Clang and LLVM, so if we use the same procedure there's less risk of "works on my machine". And improvements should always benefit both camps.

I don't know if configuration options will flow in different ways in-tree vs out-of-tree (e.g. use dylib, debug vs release, asserts vs not), but it feels like everything will be easier if there's a single configuration to develop and test.

Your patch looks nice, but it introduces a testing problem -- whenever we make a CMake change, we need to build both in-tree and out-of-tree to make sure both work. Then again, maybe that's true for build-against-installed vs build-against-build-root too.

kimgr commented on CMakeLists.txt:
```diff
@@ -97,102 +35,71 @@ add_executable(include-what-you-use
-  # Disable RTTI, use C++11 to be compatible with LLVM/Clang libraries.
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-rtti -std=c++11")
+  message(STATUS "IWYU: Linking LLVM statically")
+  set(IWYU_LLVM_LIBS LLVMSupport)
```
Thanks, I was wondering why they weren't necessary all of a sudden :)

kimgr commented:
Updated with `LLVMX86AsmParser` if we're linking statically.

kimgr commented:
Yes, it should be easy to do as you've demonstrated. I'm just concerned that the in-tree setup leaks a lot of Clang/LLVM details into our CMakeLists.txt that isn't available in the out-of-tree setup, and that keeping both working will require discipline and effort. I don't have any data whatsoever to back this up, though, so it might be worth a shot :)

Maybe we should refrain from documenting the in-tree build at least, so as to make it clear it's not offically supported?

nilsdeppe commented:
Okay, I copied the CMakeLists.txt to the clang-6 branch and can confirm this works on my Arch Linux setup. Thanks for the hard work @kimgr :)

kimgr commented:
@jru Added back support for in-tree, thanks for proving it was trivial!

jru commented:
Thank you @kim! Working on Windows too.

Dushistov commented:
Works for me, while master and clang-6.0 branch failed because of #499 ,
but I have to fix compilation error:

```diff
diff --git a/iwyu_preprocessor.cc b/iwyu_preprocessor.cc
index 91ea106..581b666 100644
--- a/iwyu_preprocessor.cc
+++ b/iwyu_preprocessor.cc
@@ -654,8 +654,8 @@ void IwyuPreprocessorInfo::InclusionDirective(
     const FileEntry* file,
     StringRef search_path,
     StringRef relative_path,
-    const clang::Module* imported,
-    SrcMgr::CharacteristicKind file_type) {
+    const clang::Module* imported/*,
+    SrcMgr::CharacteristicKind file_type*/) {
   include_filename_loc_ = filename_range.getBegin();
 }
 
diff --git a/iwyu_preprocessor.h b/iwyu_preprocessor.h
index 85c6e98..471e1a1 100644
--- a/iwyu_preprocessor.h
+++ b/iwyu_preprocessor.h
@@ -199,8 +199,8 @@ class IwyuPreprocessorInfo : public clang::PPCallbacks,
                           const clang::FileEntry* file,
                           llvm::StringRef search_path,
                           llvm::StringRef relative_path,
-                          const clang::Module* imported,
-                          clang::SrcMgr::CharacteristicKind file_type) override;
+                          const clang::Module* imported/*,
+clang::SrcMgr::CharacteristicKind file_type*/) override;
 
   void FileChanged(clang::SourceLocation loc, FileChangeReason reason,
                    clang::SrcMgr::CharacteristicKind file_type,
```

TheErk commented:
Sorry for the big lag, I'll give this a try today in our environment

TheErk commented:
Hi there,

I'm testing plain-cmake branch against llvm+clang trunk (because I want most up to date xray) and I get a link error:

```
CMakeFiles/include-what-you-use.dir/iwyu.cc.o:(.rodata._ZTIN20include_what_you_use10IwyuActionE[_ZTIN20include_what_you_use10IwyuActionE]+0x10): undefined reference to `typeinfo for clang::ASTFrontendAction'
CMakeFiles/include-what-you-use.dir/iwyu.cc.o:(.rodata._ZTIN20include_what_you_use15IwyuAstConsumerE[_ZTIN20include_what_you_use15IwyuAstConsumerE]+0x18): undefined reference to `typeinfo for clang::ASTConsumer'
CMakeFiles/include-what-you-use.dir/iwyu_output.cc.o:(.rodata._ZTIN20include_what_you_use8internal13FakeNamedDeclE[_ZTIN20include_what_you_use8internal13FakeNamedDeclE]+0x10): undefined reference to `typeinfo for clang::NamedDecl'
CMakeFiles/include-what-you-use.dir/iwyu_preprocessor.cc.o:(.rodata._ZTIN20include_what_you_use20IwyuPreprocessorInfoE+0x18): undefined reference to `typeinfo for clang::PPCallbacks'
CMakeFiles/include-what-you-use.dir/iwyu_preprocessor.cc.o:(.rodata._ZTIN20include_what_you_use20IwyuPreprocessorInfoE+0x28): undefined reference to `typeinfo for clang::CommentHandler'
```
may be it's not IWYU's fault but I wonder what can cause this?

I didn't have @Dushistov error though...

TheErk commented:
My environnement is some how special:
1) take bare debian jessie
2) boostrap/compile gcc 6.3 and install it
3) compile cmake 3.10 and install it
3) compile llvm+clang trunk with gcc and install it
4) compile iwyu and install it.

The link error comes at step 4).
The custom compile of LLVM+clang has some options:

-DCMAKE_BUILD_TYPE=MinSizeRel
-DLLVM_TARGETS_TO_BUILD="X86" 
-DGCC_INSTALL_PREFIX=<custom_prefix>/${GCC_VERSION} \
-DLLVM_ENABLE_CXX1Y=On 
-DLLVM_LINK_LLVM_DYLIB=On \
-DCMAKE_INSTALL_PREFIX:PATH=<custom_prefix>/${MY_IWYU_VERSION}/
-DLIBOMP_TSAN_SUPPORT=TRUE

TheErk commented:
Could it be that some "non pure-virtual function" in IwyuPreprocessorInfo are not implemented and trigger this ?

kimgr commented:
@Dushistov IWYU master only works with Clang trunk, so this is expected.

kimgr commented:
@TheErk Thanks for testing! Yes, I think this is something particular to your environment.

I don't think we should block this PR on this, but then maybe you'll need to patch CMakeLists.txt locally for this to work in your environment.

Are you passing the same CMake `-D` switches to the IWYU configure line? My suspicion is that IWYU somehow builds with `-frtti` -- could you see if the buildsystem generated for IWYU contains `-f.*rtti`?

TheErk commented:
By the way don't bother to take much time to try to replicate my environment. I'll do that. My env is really custom and I may not describe it all. The only things is since we build a lot of interdependant things (gcc , cmake and clang being part of the picture) reproducing the issue takes time...

Thank you for your concern & support.

kimgr commented:
> No I am not passing the same -D switches to IWYU compilation, I did expect the
> exported config from LLVM+Clang compil to be passed through on their own.

I think that's true on a flag-by-flag basis (depending on if they make it to the cache or not [I don't know the details of how CMake works, so this could be a misunderstanding]), but not sure if it holds for all of them.

kimgr commented:
One thing that stands out with your flags is `-DLLVM_LINK_LLVM_DYLIB=On`, but the lack of `-DLLVM_BUILD_LLVM_DYLIB=On`. I couldn't get the LLVM build on my FreeBSD system to link. Trying with BUILD_LLVM_DYLIB now to see how things work out.

kimgr commented:
Yes! I can repro if I add `-DLLVM_BUILD_LLVM_DYLIB`. Now to try and find out why :)

kimgr commented:
I found that `add_llvm_executable` is what applies the compiler flags, and I had reverted to just use `add_executable`. I've rewritten the PR branch to use `add_llvm_executable` again. Please give it another shot!

TheErk commented:
I'll do that (on monday though) thanks you for investigating.


Le ven. 25 mai 2018 07:43, Kim Gräsman <notifications@github.com> a écrit :

> I found that add_llvm_executable is what applies the compiler flags, and
> I had reverted to just use add_executable. I've rewritten the PR branch
> to use add_llvm_executable again. Please give it another shot!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/include-what-you-use/include-what-you-use/pull/542#issuecomment-391947733>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAVvwF-xcLt3XXC4ndxINlrWe35Eldo9ks5t15n8gaJpZM4T8hVq>
> .
>

kimgr commented:
I pushed some updates to the documentation, hopefully to make it clearer. If you've reviewed, it might be worth taking another look.

TheErk commented:
Hi Kim,
I gave the latest tip of plain-cmake branch a try and it's working for me.
Thanks you for this. AFAI am concerned this PR is good to go for merging to master.

Thanks again.
Eric
